pkg/cloudprovider/vsphere/nodemanager_test.go

```diff
@@ -267,3 +268,825 @@ func TestReturnIPsFromSpecificFamily(t *testing.T) {
+						IpAddress: []string{
+							"127.0.0.7",
+							"172.15.108.10",
+							"172.15.108.11",
+						},
+					},
+				},
+			},
+			expectedIPs: []v1.NodeAddress{
+				{Type: "InternalIP", Address: "10.10.1.22"},
+				{Type: "ExternalIP", Address: "10.10.1.22"},
+			},
+		},
+		{
+			testName: "ByNetwork_whenOnlyInternalNetworkIsSet_itReturnsOnlyInternalIP",
+			setup: testSetup{
+				ipFamilyPriority: []string{"ipv4"},
+				cpiConfig: &ccfg.CPIConfig{
+					Nodes: ccfg.Nodes{
+						InternalVMNetworkName: "test_k8s_tenant_c123_internal",
+					},
+				},
+				networks: []vimtypes.GuestNicInfo{
+					{
+						Network: "test_k8s_tenant_c123_internal",
+						IpAddress: []string{
+							"127.0.0.6",
+							"10.10.1.22",
+							"10.10.1.23",
+						},
+					},
+					{
+						Network: "test_k8s_tenant_c123_external",
+						IpAddress: []string{
+							"127.0.0.7",
+							"172.15.108.10",
+							"172.15.108.11",
+						},
+					},
+				},
+			},
+			expectedIPs: []v1.NodeAddress{
+				{Type: "InternalIP", Address: "10.10.1.22"},
+			},
+		},
+		{
+			testName: "BySubnetAndNetworkNameTwoNICs_desiredIPsAreSplitAcrossNICs",
+			setup: testSetup{
+				ipFamilyPriority: []string{"ipv4"},
+				cpiConfig: &ccfg.CPIConfig{
+					Nodes: ccfg.Nodes{
+						InternalNetworkSubnetCIDR: "10.10.0.0/16",
+						ExternalVMNetworkName:     "test_another_nic",
+					},
+				},
+				networks: []vimtypes.GuestNicInfo{
+					{
+						Network: "test_k8s_tenant_c123",
+						IpAddress: []string{
+							"127.0.0.6",
+							"169.0.1.2",
+							"10.10.1.22",
+						},
+					},
+					{
+						Network: "test_another_nic",
+						IpAddress: []string{
+							"127.0.0.7",
+							"172.15.108.11",
+						},
+					},
+				},
+			},
+			expectedIPs: []v1.NodeAddress{
+				{Type: "InternalIP", Address: "10.10.1.22"},
+				{Type: "ExternalIP", Address: "172.15.108.11"},
+			},
+		},
+		{
+			// This seems like undesireable behavior. The IPs that are returned
+			// depend on ordering within the NICs or which NIC they are on.
+			// There is no sensible precedence with the configuration.
+			testName: "BySettingBothNetworkNameAndSubnets_precedenceIsNotObvious",
+			setup: testSetup{
+				ipFamilyPriority: []string{"ipv4"},
+				cpiConfig: &ccfg.CPIConfig{
+					Nodes: ccfg.Nodes{
+						InternalNetworkSubnetCIDR: "10.10.0.0/16",
+						ExternalNetworkSubnetCIDR: "172.15.0.0/16",
+						InternalVMNetworkName:     "internal_net",
+						ExternalVMNetworkName:     "external_net",
+					},
+				},
+				networks: []vimtypes.GuestNicInfo{
+					{
+						Network: "internal_net",
+						IpAddress: []string{
+							"22.22.22.22",
+							"172.15.108.11",
+						},
+					},
+					{
+						Network: "external_net",
+						IpAddress: []string{
+							"33.33.33.33",
+							"10.10.1.22",
+						},
+					},
+				},
+			},
+			expectedIPs: []v1.NodeAddress{
+				{Type: "InternalIP", Address: "22.22.22.22"},
+				{Type: "ExternalIP", Address: "172.15.108.11"},
+			},
```

Comment:
This seems like undesireable behavior. The IPs that are returned
depend on ordering within the NICs or which NIC they are on.
There is no sensible precedence with the configuration.

Reply:
I think defining the precedence here is warranted.

Reply:
I feel like there's a bug in the code - I think the ExternalIP is expected as we prioritize address masking over network name. However, for InternalIP, it looks like the network name has higher priority, which does not look right to me

Reply:
I agree, this is incorrect behavior. I'd like to keep this PR as merely test coverage (noting this expectation is undesireable), and then follow up with a PR to refactor and correct these problems. The refactor PR diff will highlight the bug fixes/changes in behavior in the test.